File: _build_artifact.py
import re, base64, urllib.request, sys
UA="Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36"
css=open('_fonts.css','r',encoding='utf-8').read()
urls=sorted(set(re.findall(r'https://[^)\s]+\.woff2',css)))
print('fonts to inline:',len(urls))
cache={}
for u in urls:
    req=urllib.request.Request(u,headers={'User-Agent':UA})
    data=urllib.request.urlopen(req,timeout=30).read()
    b64=base64.b64encode(data).decode()
    cache[u]='data:font/woff2;base64,'+b64
    print('  ok',u.split('/')[-1],len(data),'bytes')
for u,d in cache.items():
    css=css.replace(u,d)
# build mockup artifact: strip outer wrappers + google fonts <link>, prepend inlined @font-face
mk=open('mockup.html','r',encoding='utf-8').read()
# remove google fonts links + preconnect
mk=re.sub(r'<link[^>]*fonts\.(googleapis|gstatic)[^>]*>','',mk)
# strip doctype/html/head/body wrapper tags (keep inner content)
mk=re.sub(r'<!doctype html>','',mk,flags=re.I)
mk=re.sub(r'</?html[^>]*>','',mk,flags=re.I)
mk=re.sub(r'</?head[^>]*>','',mk,flags=re.I)
mk=re.sub(r'</?body[^>]*>','',mk,flags=re.I)
out='<style>\n'+css+'\n</style>\n'+mk
open('mockup-artifact.html','w',encoding='utf-8').write(out)
print('WROTE mockup-artifact.html bytes:',len(out))

File: _build_artifact.js
const fs=require('fs'), https=require('https');
const UA='Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0 Safari/537.36';
function get(url){return new Promise((res,rej)=>{https.get(url,{headers:{'User-Agent':UA}},r=>{const c=[];r.on('data',d=>c.push(d));r.on('end',()=>res(Buffer.concat(c)));}).on('error',rej);});}
(async()=>{
  let css=fs.readFileSync('_fonts.css','utf8');
  const urls=[...new Set(css.match(/https:\/\/[^)\s]+\.woff2/g)||[])];
  console.log('fonts to inline:',urls.length);
  for(const u of urls){ const buf=await get(u); css=css.split(u).join('data:font/woff2;base64,'+buf.toString('base64')); console.log('  ok',u.split('/').pop(),buf.length,'bytes'); }
  let mk=fs.readFileSync('mockup.html','utf8');
  mk=mk.replace(/<link[^>]*fonts\.(googleapis|gstatic)[^>]*>/gi,'');
  mk=mk.replace(/<!doctype html>/i,'').replace(/<\/?html[^>]*>/gi,'').replace(/<\/?head[^>]*>/gi,'').replace(/<\/?body[^>]*>/gi,'');
  const out='<style>\n'+css+'\n</style>\n'+mk;
  fs.writeFileSync('mockup-artifact.html',out);
  console.log('WROTE mockup-artifact.html bytes:',out.length);
})();
